lib/phoenix_live_view/js.ex

```diff
@@ -218,10 +218,89 @@ defmodule Phoenix.LiveView.JS do
 
   defimpl Phoenix.HTML.Safe, for: Phoenix.LiveView.JS do
     def to_iodata(%Phoenix.LiveView.JS{} = js) do
-      Phoenix.HTML.Engine.html_escape(Phoenix.json_library().encode!(js.ops))
+      js
+      |> JS.encode()
+      |> Phoenix.HTML.Engine.html_escape()
     end
   end
 
+  @doc ~S"""
+  Encodes the JS commands for transmission to the client.
+
+  Most of the time you will not need to call this function directly. JS commands
+  are typically rendered in [HEEx templates](assigns-eex.md), in which they are
+  encoded automatically.
+
+  This function is useful when constructing JS commands dynamically on the
+  server and pushing them to the client via `Phoenix.LiveView.push_event/3`.
+
+  ## Examples
+
+  On the server, dynamically compute some JS commands and push them to the
+  client:
+
+  ```elixir
+  socket
+  |> push_event("myapp:exec_js", %{
+    to: "#items-#{item.id}",
+    js: js_commands_for(item) |> JS.encode()
+  })
+  ```
+
+  On the client, handle the event and execute the commands:
+
+  ```javascript
+  window.addEventListener("phx:myapp:exec_js", e => {
+    const {to, js} = e.detail;
+    const el = document.querySelector(to);
+    if (el && js) {
+      window.liveSocket.execJS(el, js);
+    }
+  });
+  ```
+
+  The common case, though, is having the JS commands stored in an HTML attribute
+  (`phx-*` or `data-*`), such that client-side JavaScript can refer to them
+  later. For example, in a LiveView template:
+
+  ```heex
+  <div id={"items-#{item.id}"} data-js={JS.show()}>
+    Hello!
+  </div>
+  ```
+
+  Now the server can push an event that refers to the `data-js` attribute:
+
+  ```elixir
+  socket
+  |> push_event("myapp:exec_attr", %{
+    to: "#items-#{item.id}",
+    attr: "data-js"
+  })
+  ```
+
+  Finally, on the client, you can read and execute the commands:
+
+  ```javascript
+  window.addEventListener("phx:myapp:exec_attr", e => {
+    const {to, attr} = e.detail;
+    const el = document.querySelector(to);
+    const js = el && attr && el.getAttribute(attr);
+    if (el && js) {
+      window.liveSocket.execJS(el, js);
+    }
+  });
+  ```
+
+  Note how in the code above we didn't need to encode JS commands explicitly,
+  nor to pass them in the event payload, thanks to rendering them in the
+  template.
+
+  """
+  def encode(%JS{} = js) do
+    Phoenix.json_library().encode!(js.ops)
```

Comment:
It is probably better to just make it return `js.ops`, so you can arbitrarily nest it elsewhere. Maybe we need to change the name a bit (`encodable`?) but it will probably be better if we don't prescribe the particular encoding.

Reply:
Let's go with js.ops, but mention in the docs that the return value is opaque and the only guarantee we give is that is is JSON-encodeable. We'll also need to add a check in `js.js`'s exec function and skip JSON.parse in case it's already an Array.

Reply:
On the client side, "encoded JS" must be a JSON-encoded array:

https://github.com/phoenixframework/phoenix_live_view/blob/f4344b4a020e55dfe33694439d269892d8af475c/assets/js/phoenix_live_view/js.js#L7-L17

So the serialization format is not something users can choose.

An opaque binary is less API surface than exposing the `JS.ops` list, isn't it? Nesting/combining commands would have to happen either before calling `encode`, or arranging to `exec` multiple "encoded JS" on the client.

Reply:
Correct, there's nothing to be combined by users, it should be considered opaque. José's point is that double encoding is minimally more expensive to do, and we can omit it for this case.

Reply:
I wouldn't do anything about the unstable sort in this PR. I don't have a good solution, but I'm also not sure if it's really big problem.